expressions/literals.md

```diff
@@ -0,0 +1,319 @@
+Array literal elements can be any kind of expressions.
+They are separated by semicolon or newline:
+
+```pony
+let my_literal_array = [ "first"; "second"
+  "third one on a new line"]
+```
+
+Empty array literals are not supported.
+Use ``recover val Array(0) end`` instead.
+
+### Type inference
+
+
+The type of the literal array expression is Always ``Array[T] ref``
+where ``T`` (the type of the elements) is inferred as 
+the union type of all the element types:
+
+```pony
+let my_heterogenous_array: Array[(U64|String)] ref = [
+  U64(42)
+  "42"
+  U64.min_value()
+]
+```
+
+In the above example the resulting array type will be
+``Array[(U64|String)] ref`` because the array contains
+``String`` and ``U64`` elements.
+
+It is possible to give the literal a hint on what kind of type it should
+coerce the array literal to. In the above example
+we might only want an ``Array[Stringable] ref``. ``Stringable`` is a trait
+that both ``String`` and ``U64`` implement.
+
+In order to do so, add an ``as`` Expression
+defining the desired Array element type right after the opening square bracket
+delimited by a colon:
+
+```pony
+let my_stringable_array: Array[Stringable] ref =
+  [as Stringable:
+    U64(0xFFEF)
+    "0xFFEF"
+    U64(1 + 1)
+  ]
+```
+
+This Array literal is coerced to be an ``Array[Stringable] ref`` according to
+the ``as`` expression.
+
+Inferring the element type from the type of the local variable
+or field it is assigned to is not yet possible:
+
+```pony
+// does not compile
+let my_stringable_array: Array[Stringable] =
+  [
+    U64(0xA)
+    "0xA"
+  ]
+```
+
+### Arrays and References
+
+*This section will lead you down the rabbit-hole of
+[Reference capabilities](../capabilities/index.md).
+It is advised to make yourself familiar with that concept first before reading this.*
+
+To be 100% technically correct, array literal elements are inferred
+to be the alias of the the actual element type.
+If all elements are of type ``T`` the array literal will be inferred as
+``Array[T!]`` that is as an array of aliases of the type ``T``.
+
+***Why is that?***
+
+Arrays store references to their elements, not the elements themselves.
+There might exist another reference somewhere else to an array element.
+In order to safely use ``Array`` with elements of all types and
+reference capabilities (including ``iso``), the only possibility is to store
+the elements as an alias to that type: ``T!``
+
+To safely create an array literal of type ``Array[T]`` it is necessary
+to use elements of type ``T^``, an ephemeral type of ``T``;
+A ``T`` that is not assigned to any variable yet.
+"Consuming" a variable returns an ephemeral type,
+all constructors return ephemeral types, and other literal expressions work
+the same:
+
+```pony
+let ref = U64(123)
+let safe_array: Array[U64 val] =
+  [
+    consume ref     // consuming a variable
+    U64.create(1)   // constructor
+    U64.min_value() // constructor
+  ]
```

Comment:
Your array style is inconsistent throughout this page. Please check the style guide in the ponyc repo for the style that is used in the standard library and should be adopted across documentation as well.